### PR TITLE
Create initial linting rules

### PR DESCRIPTION
**Pull request by beanc16**
This PR creates the initial linting rules for this package and installs any necessary dependencies. All linting rules should have a comment next to them explaining roughly what they do.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,5 +20,23 @@
...
+        "eslint-import-resolver-typescript": "3.6.1",
+        "eslint-plugin-destructuring-newline": "0.0.2",
+        "eslint-plugin-etc": "2.0.3",
+        "eslint-plugin-file-progress": "1.3.0",
+        "eslint-plugin-import": "2.28.1",
+        "eslint-plugin-import-newlines": "1.3.4",
+        "eslint-plugin-jest": "27.4.2",
+        "eslint-plugin-simple-import-sort": "10.0.0",
+        "typescript-eslint": "8.7.0"
+    }
```
Comment (beanc16) [inline package.json]: I'm uncertain if these need to be dependencies, peer dependencies, or dev dependencies (or multiple). Will triaging that locally once the rules are fleshed out more to confirm.

File: .eslintrc.cjs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            ],
+            extends: [
+                'airbnb-typescript/base',
+                'plugin:@stylistic/recommended-extends',
+                'plugin:@stylistic/disable-legacy',
+            ],
+            parser: '@typescript-eslint/parser',
+            parserOptions: {
+                project: [
+                    path.resolve(path.resolve(), 'tsconfig.json'),          // Required for: @typescript-eslint/recommended-requiring-type-checking
```
Comment (beanc16) [inline .eslintrc.cjs:143]: I'd like to include `@typescript-eslint/recommended-requiring-type-checking`, though I haven't yet, as I'm not sure if it's necessary. Will revisit before this PR is out of draft state.

File: package.json (hunk 1 above)
Comment (beanc16) [reply package.json]: Upon testing locally, plugins need to be peer dependencies according to eslint documentation. Additionally, I found that `eslint-import-resolver-typescript` needed to be a peer dependency as well, otherwise errors like this would get spammed to the console in every TypeScript file:
```
1:1   error  Resolve error: typescript with invalid interface loaded as resolver      import/no-relative-packages
  1:27  error  Unexpected use of file extension "js" for "./path-to-your-file.js"  import/extensions
```

File: .eslintrc.cjs (hunk 2 above)
Comment (beanc16) [reply .eslintrc.cjs:143]: This has been included with one small modification to the default settings to make promise handling smoother
https://github.com/nick-w-nick/eslint-config-lintification/pull/1/commits/f9b27eec34c466c655395d6ca0bbd5be70e78827

File: README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+
+
+## Table of Contents
+- [eslint-config-lintification](#eslint-config-lintification)
+  - [Table of Contents](#table-of-contents)
+  - [Usage](#usage)
+    - [Install](#install)
+    - [Extend the Configuration](#extend-the-configuration)
+  - [License](#license)
```
Comment (beanc16) [inline README.md]: I included the table of contents out of habit. But since it's such a short README, it's probably not necessary.
```suggestion
```

File: package.json [hunk 4]
```diff
@@ -18,7 +18,41 @@
...
+    "peerDependencies": {
+        "@stylistic/eslint-plugin": "2.10.1",
+        "eslint": ">= 8.56.0",
+        "eslint-plugin-file-progress": "1.3.0",
+        "eslint-plugin-import": "2.28.1",
+        "eslint-plugin-import-newlines": "1.3.4",
+        "eslint-plugin-jest": "27.4.2",
+        "eslint-plugin-newline-destructuring": "1.2.2",
+        "eslint-plugin-simple-import-sort": "10.0.0"
+    }
```
Comment (nick-w-nick) [inline package.json]: Can you summarize why some have to be peer dependencies and others not?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+### Extend the Configuration
+
+Either add `"extends": "lintification"` to your linting configuration or add the below to your package.json to use the rules with no modifications:
+```json
+{
+    "eslintConfig": {
+        "extends": "lintification"
+    }
+}
+```
```
Comment (nick-w-nick) [inline README.md]: Can you reorganize the sections to be the following?
- How to install and add to package.json (should just be `npm install eslint-config-lintification` and then add `eslintConfig` to package.json)
- How to extend/overwrite rules with an example for how to add/disable/change rules
- How to contribute to the package (shouldn't need to install all of those deps manually, they should be able to be marked as dev dependencies in the package.json)

File: .eslintrc.cjs
Comment (nick-w-nick) [inline .eslintrc.cjs:1]: @beanc16 Any reason why this file can't be `eslintrc.js`?

File: package.json (hunk 4 above)
Comment (beanc16) [reply package.json]: That's summarized in this thread: https://github.com/nick-w-nick/eslint-config-lintification/pull/1#discussion_r1845742004

File: README.md (hunk 5 above)
Comment (beanc16) [reply README.md]: > How to install and add to package.json (should just be npm install eslint-config-lintification and then add eslintConfig to package.json)

This isn't necessarily true. Peer dependencies need to be installed as well in order for things to work (at least when testing locally with `npm link`). `npm link` has led to differences with linting rules packages like this before though, so I'd be willing to go back and edit that once it's ready for consumption for the first time if the peers are only necessary for local linking.

Comment (beanc16) [reply README.md]: Could you also propose a suggestion for what you're wanting out of this documentation? I'm unclear on what you actually want changed. Using [airbnb-base's README](https://www.npmjs.com/package/eslint-config-airbnb-base) as an example, I opted not to include instructions on how to override rules, since that's something that can easily be searched outside the confines of this specific linting config in eslint's own documentation.

File: .eslintrc.cjs
Comment (beanc16) [reply .eslintrc.cjs:1]: I saw this popup in VSCode.
image

Also didn't see any reason not to just have it be `.cjs` since that does the job and is truer to the current config.

File: package.json (hunk 4 above)
Comment (nick-w-nick) [reply package.json]: @beanc16 I saw, but that doesn't say why for example `eslint-import-resolver-typescript` is a dependency but not a peer dependency, while `@stylistic/eslint-plugin` is both.

Comment (beanc16) [reply package.json]: This part is most important: `plugins need to be peer dependencies`. When doing `npm link` locally, if something is a plugin, I needed it as a peer dependency and to be installed as a dev dependency in the consuming project, otherwise, linting wouldn't work and would say the plugin was missing. If it's not a plugin, I found it could be a regular dependency and not installed in the consuming project.

The one exception I found to that rule of thumb was `eslint-import-resolver-typescript`, which threw these errors in every file if it wasn't listed as a peer dependency:
```
1:1   error  Resolve error: typescript with invalid interface loaded as resolver      import/no-relative-packages
  1:27  error  Unexpected use of file extension "js" for "./path-to-your-file.js"  import/extensions
```

Though, what's strange is [eslint's own documentation](https://eslint.org/docs/latest/extend/shareable-configs#publishing-a-shareable-config) says to list it as a `dependency`, so it's unclear to me if this is just strange behavior coming from how `npm link` intersects with linting config packages or not.

File: README.md (hunk 5 above)
Comment (nick-w-nick) [reply README.md]: @beanc16 Peer dependencies don't need to be installed manually, they should be automatically installed when you install the package.

For the readme, something like this:
````suggestion
### Installation
`npm install eslint-config-lintification`

If you have an existing `eslint.config.js` file, add it to your list of plugins as defined [here](https://eslint.org/docs/latest/extend/shareable-configs#using-a-shareable-config) in the eslint documentation/

If you do not have a pre-existing configuration file, you can simply add the configuration to your `package.json` via the `eslintConfig` property:
```json
{
    "eslintConfig": {
        "extends": "eslint-config-lintification"
    }
}
```

Finally, ensure you have a valid linting script added to your `package.json`, like so:
```json
{
    "scripts": {
        "lint": "eslint --cache --ext js --ext ts ./",
        "lint:fix": "eslint --cache --fix --ext js --ext ts ./"
    }
}
```

Then, just run `npm run lint` to run the linter, and run `npm run lint:fix` to fix any issues that can automatically be fixed by the linter.

### Extend the Configuration
You can easily add new or overwrite existing rules by updating your configuration file or in the `eslintConfig` property of your `package.json`.

If you have a pre-existing `eslint.config.js` file, see [this page](https://eslint.org/docs/latest/extend/shareable-configs#overriding-settings-from-shareable-configs) from the eslint documentation on extending shared configurations.

If you don't use a traditional configuration file, you can still extend the configuration by including them in the `rules` property of the `eslintConfig` object in your `package.json`.
```json
{
    "eslintConfig": {
        "extends": "eslint-config-lintification",
        "rules": {
            "object-curly-spacing": "off"
        }
    }
}
```
````

Not worried about the contributing section yet, we can add that later.

File: .eslintrc.cjs
Comment (nick-w-nick) [reply .eslintrc.cjs:1]: @beanc16 My only worry is still being able to support ESM imports. Can you test to make sure this type of import still works?

https://eslint.org/docs/latest/extend/shareable-configs#overriding-settings-from-shareable-configs

Comment (nick-w-nick) [reply .eslintrc.cjs:1]: @beanc16 Actually, looking at the documentation, `eslintrc.js`/`eslintrc.cjs` is the deprecated way of doing it:
image

We are supposed to be using `eslint.config.js` now with ESM and default exports according to [this page](https://eslint.org/docs/latest/extend/shareable-configs#creating-a-shareable-config).
Can you make those changes?

File: README.md (hunk 5 above)
Comment (beanc16) [reply README.md]: > Peer dependencies don't need to be installed manually, they should be automatically installed when you install the package.

From my research, seems like what I encountered is just an issue with `npm link` and peer dependencies. We should figure out the simplest way to link locally for maintenance and contribution documentation in the future.

One thing I've found is a `--resolve-plugins-relative-to` flag that'd be used similar to this:
```
        "lint": "eslint --resolve-plugins-relative-to './path-on-your-local-machine-to/eslint-config-lintification' --cache --ext js --ext ts ./",
```

Though that's also imperfect and throws this error:
```
Error: Error while loading rule 'jest/no-deprecated-functions': Unable to detect Jest version - please ensure jest package is installed, or otherwise set version explicitly
```

File: package.json (hunk 4 above)
Comment (nick-w-nick) [reply package.json]: The documentation says to list `eslint` as a peer dependency (red), and make the plugins regular dependencies (green).

Screenshot 2024-11-18 at 12 07 09 PM

I can't totally tell if it means they need to be dependencies in the consuming project's package.json or if it means in ours, but it might work with just eslint in peer deps and the rest in regular deps.

File: README.md (hunk 5 above)
Comment (nick-w-nick) [reply README.md]: @beanc16 Try the changes from this comment and see if it still works properly: https://github.com/nick-w-nick/eslint-config-lintification/pull/1#discussion_r1846964202

File: .eslintrc.cjs
Comment (beanc16) [reply .eslintrc.cjs:1]: Updated

File: README.md (hunk 5 above)
Comment (beanc16) [reply README.md]: It's no better nor worse, the `jest/no-deprecated-functions` error still occurs.

Comment (beanc16) [reply README.md]: I applied the changes to the README with some notable changes:
1. Added `-D` to the suggestion so that it'd be installed as a dev dependency in consuming projects.
2. All `extends` fields should evaluate to `lintification`, not `eslint-config-lintification`. You'll notice we don't do that for any other config we extend either, as that's how eslint processes it. https://github.com/nick-w-nick/eslint-config-lintification/blob/c459ec9f0fd1a30012dc389b579c5b808dab2f68/eslint.config.js#L8

Comment (beanc16) [reply README.md]: I was able to resolve this error by adding `jest` as a dev dependency to this package

Comment (nick-w-nick) [reply README.md]: @beanc16 How does that work for when the package gets consumed though since the dev dependencies will be removed? Shouldn't Jest just be an optional peer dependency?

File: eslint.config.js
Comment (nick-w-nick) [inline eslint.config.js:1]: @beanc16 Can you make this file perform a default export instead, as defined on [this page](https://eslint.org/docs/latest/extend/shareable-configs#creating-a-shareable-config)? You likely need to set `"type": "module"` in the `package.json` as well.

File: README.md (hunk 5 above)
Comment (beanc16) [reply README.md]: Jest is due to the `--resolve-plugins-relative-to` flag. It only really matters when linking locally with `npm link`. Otherwise, it doesn't get used.

It fixes the below error, presumably by having jest in the correct place that it's checking for it with that plugin (again, only when linking locally).
```
Error: Error while loading rule 'jest/no-deprecated-functions': Unable to detect Jest version - please ensure jest package is installed, or otherwise set version explicitly
```

File: eslint.config.js
Comment (beanc16) [reply eslint.config.js:1]: We discussed over discord VC and agreed to keep the legacy style config for the time being and potentially migrate to a flat config in the future.

**Comment by beanc16** [comment]
@nick-w-nick Can I get a re-review of this when you get the chance?

File: .eslintrc.cjs [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+                'airbnb-typescript/base',
+                'plugin:@stylistic/recommended-extends',
+                'plugin:@stylistic/disable-legacy',
+                'plugin:@typescript-eslint/recommended-requiring-type-checking',
+            ],
+            parser: '@typescript-eslint/parser',
+            parserOptions: {
+                project: [
+                    path.resolve(path.resolve(), 'tsconfig.json'),          // Required for: @typescript-eslint/recommended-requiring-type-checking
+                    path.resolve(path.resolve(), 'tests/tsconfig.json'),    // Required for the below jest overrides to take effect for typescript files
```
Comment (nick-w-nick) [inline .eslintrc.cjs:145]: @beanc16 Does this mean that anyone using this package needs to have their tests in a folder called `tests`? I've seen some projects just call them `test`. If that is the case, we will either need a way to modify/override this easily on the consuming side.

File: package.json [hunk 7]
```diff
@@ -18,7 +18,37 @@
         "url": "git+https://github.com/nick-w-nick/eslint-config-lintification.git"
     },
     "homepage": "https://github.com/nick-w-nick/eslint-config-lintification#readme",
-    "keywords": ["eslint", "lintification", "lint", "typescript", "nodejs", "backend"],
+    "keywords": [
+        "eslint",
```
Comment (nick-w-nick) [inline package.json:22]: @beanc16 You think these are worth adding?

```suggestion
        "eslint",
        "eslint-config",
        "config",
```

Comment (beanc16) [reply package.json:22]: Perhaps with config omitted if you'd like?
```suggestion
        "eslint",
        "eslint-config",
```

File: .eslintrc.cjs (hunk 6 above)
Comment (beanc16) [reply .eslintrc.cjs:145]: I did some thinking on this. This was originally added because it assumes your main `tsconfig.json` doesn't include your test file like so:
```json
{
    ...
    "include": [
        "src",
    ],
    ...
}
```

And then assumes you have a second `tsconfig.json` for your unit tests that's something like this:
```json
{
    ...
    "extends": "../tsconfig.json",
    "include": [
        "**/*.test.ts", // or "**/*.spec.ts", whatever the extension is for their Jest setup
    ],
    ...
}
```

Ultimately, what matters is that your tsconfig includes your test files. With that in mind, I think this would cover the most cases most of the time:
```json
            parserOptions: {
                project: [
                    path.resolve(path.resolve(), 'tsconfig.json'),          // Required for: @typescript-eslint/recommended-requiring-type-checking
                    path.resolve(path.resolve(), 'test/tsconfig.json'),     // Required for the below jest overrides to take effect for typescript files
                    path.resolve(path.resolve(), 'tests/tsconfig.json'),    // Required for the below jest overrides to take effect for typescript files
                ],
            },
```

Comment (beanc16) [reply .eslintrc.cjs:145]: Updated

File: package.json (hunk 7 above)
Comment (nick-w-nick) [reply package.json:22]: Works for me!

File: .eslintrc.cjs [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+            ],
+            extends: [
+                'plugin:jest/recommended',
+            ],
+            parser: '@typescript-eslint/parser',
+            rules: {
+                '@typescript-eslint/dot-notation': 'off', // Allow for object['property'] notation for testing of protected/private classes
+            },
+        },
+    ],
```
Comment (nick-w-nick) [inline .eslintrc.cjs:251]: Can you add a TODO here about possibly adding support for other testing frameworks such as mocha or vitest?

**Review by nick-w-nick — APPROVED** [review]
Approved barring one last minor suggestion

File: .eslintrc.cjs (hunk 8 above)
Comment (beanc16) [reply .eslintrc.cjs:251]: I created an issue: https://github.com/nick-w-nick/eslint-config-lintification/issues/4
